harness/determined/core/_profiler.py

```diff
@@ -0,0 +1,396 @@
+            _CPU(),
+            _Memory(),
+            _Disk(),
+            _Network(),
+        ]
+        self._should_exit = threading.Event()
+
+        super().__init__(daemon=True)
+
+    def run(self) -> None:
+        collection_stop_ts = time.time() + self._max_collection_s
+        while not self._should_exit.is_set():
+            if time.time() > collection_stop_ts:
+                # Max collection time reached. Notify queue consumers and exit.
+                self._metrics_queue.put(_ShutdownMessage())
+                return
+
+            self._collect_metrics()
+            self._should_exit.wait(self._collection_interval)
+
+    def stop(self) -> None:
+        self._should_exit.set()
+        self._metrics_queue.put(_ShutdownMessage())
+
+    def _collect_metrics(self) -> None:
+        for collector in self._metric_collectors:
+            timestamp = datetime.now(timezone.utc)
+            metrics = collector.collect()
+            if not metrics:
+                continue
+            group_metrics = _Metric(group=collector.group, metrics=metrics, timestamp=timestamp)
+            self._metrics_queue.put(group_metrics)
+
+
+class _Shipper(threading.Thread):
+    def __init__(
+        self,
+        session: api.Session,
+        agent_id: str,
+        trial_id: int,
+        run_id: int,
+        metrics_queue: queue.Queue,
+        flush_interval: int = 10,
+    ):
+        self._metrics_queue = metrics_queue
+        self._should_exit = False
+        self._flush_interval = flush_interval
+        self._session = session
+        self._agent_id = agent_id
+        self._trial_id = trial_id
+        self._run_id = run_id
+
+        self._metrics: List[_Metric] = []
+
+        super().__init__(daemon=True)
+
+    def run(self) -> None:
+        while True:
+            next_flush_ts = time.time() + self._flush_interval
+            # Pop messages from queue until the next scheduled flush time.
+            while time.time() < next_flush_ts:
+                deadline = next_flush_ts - time.time()
+                try:
+                    msg = self._metrics_queue.get(timeout=deadline)
+                    if isinstance(msg, _ShutdownMessage):
+                        # Received shutdown message, ship accumulated metrics and exit.
+                        self._ship()
+                        return
+
+                    self._metrics.append(msg)
+                except queue.Empty:
+                    # Timeout reached, no messages in queue.
+                    break
+
+            self._ship()
+
+    def stop(self) -> None:
+        self._metrics_queue.put(_ShutdownMessage())
+
+    def _ship(self) -> None:
+        # xxx: change this to popleft from list?
+        for metric in self._metrics:
+            # Append agent ID to every metric reported.
+            metrics = {self._agent_id: metric.metrics}
+            self._report_metrics(group=metric.group, timestamp=metric.timestamp, metrics=metrics)
```

Comment:
wut?  if you are doing a separate HTTP request for each metric, why are you even bothering with a loop to collect batches of metrics?